### ISSUE TITLE
DevOps: Cannot export Intune production environment using CertThumbprint

### ISSUE THREAD

**Issue by ricmestre** [microsoft/Microsoft365DSC#3340]
#### Details of the scenario you tried and the problem that is occurring

Currently I have 3 tenants, 2 which are for development (both managed by a single DevOps Instance) and are working perfectly and 1 for production which uses a separate DevOps instance but has exactly the same code, in fact all 3 tenants retrieve the certificate/secret from a KeyVault in the production env.

I have a custom script on both DevOps instances that connects to Azure to retrieve the certificate/secret for Intune, which works, then I import the certificate and go straight to Export-M365DSCConfiguration which is failing, only in the production instance not on the dev ones, with the same problem as mentioned in [#1922](https://github.com/microsoftgraph/msgraph-sdk-powershell/issues/1922) which was never solved (see below). It doesn't matter if I run the task in MS-hosted pipelines or on my own Azure VM, or running with a combination of local PowerShell or AzurePowerShell tasks (which doesn't need a separate KeyVault task and therefore Connect-AzAccount won't be needed from within the script), it will always fail with the same problem and only happens on the production environment and if running on DevOps, but *runs without issues locally in the Azure VM*.

I already tried to recreate the Application and the Certificate several times, and of course re-uploading it to the KeyVault, and *bruteforcing* with the following below, but as soon as it tries to connect to M365 to run the export it fails. Also to note that I tested a separate pipeline in DevOps just calling a test script from my Azure VM with only Export-M365DSCConfiguration with hardcoded parameters, and not connecting to Azure, and that works!!! So the problem really is the fact that there is a connection to Azure prior to the export but for the life of me I cannot find a workaround for this.

Any clues?

```powershell
Disconnect-AzAccount -ErrorAction SilentlyContinue | Out-Null
Get-AzContext | %{
    Remove-AzContext -InputObject $_ -Confirm:$false -Force
}
Clear-AzContext -Scope CurrentUser -Force -ErrorAction SilentlyContinue
Clear-AzContext -Scope Process
Disconnect-Graph -ErrorAction SilentlyContinue | Out-Null
$global:MSCloudLoginConnectionProfile = $NULL
Remove-Module -Name Az.* -Force
Remove-Module -Name Microsoft.Graph.* -Force
```

#### Verbose logs showing the problem

```powershell
##[error]Connect-MgGraph : You specified a different tenant - once in WithAuthority() and once using WithTenant().
At C:\Program Files\WindowsPowerShell\Modules\MSCloudLoginAssistant\1.0.110\Workloads\MicrosoftGraph.psm1:101 char:21
+ ...             Connect-MgGraph -ClientId $Global:MSCloudLoginConnectionP ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Connect-MgGraph], MsalClientException
    + FullyQualifiedErrorId : Microsoft.Graph.PowerShell.Authentication.Cmdlets.ConnectMgGraph
##[error]PowerShell exited with code '1'.
```

#### Suggested solution to the issue

N/A

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)

N/A

#### The operating system the target node is running

OsName               : Microsoft Windows Server 2022 Datacenter Azure Edition
OsOperatingSystemSKU : 407
OsArchitecture       : 64-bit
WindowsVersion       : 2009
WindowsBuildLabEx    : 20348.1.amd64fre.fe_release.210507-1500
OsLanguage           : en-US
OsMuiLanguages       : {en-US}
PSVersionTable.PSVersion: 5.1.20348.1366

#### Version of the DSC module that was used ('dev' if using current dev branch)

1.23.517.1

**Comment by ricmestre** [comment]
Also to note, I added a try/catch around Export-M365DSCConfiguration and the variables below which are passed to Connect-MgGraph are all correct.

$Global:MSCloudLoginConnectionProfile.MicrosoftGraph.CertificateThumbprint
$Global:MSCloudLoginConnectionProfile.MicrosoftGraph.ApplicationId
$Global:MSCloudLoginConnectionProfile.MicrosoftGraph.TenantId
$Global:MSCloudLoginConnectionProfile.MicrosoftGraph.GraphEnvironment

**Comment by andikrueger** [comment]
Could you share the script used to retrieve the certificate as well? If I get it correctly, there are the following steps involved:

1. Connect to Azure 
2. Export the Certificate
3. Disconnect Azure
4. Connect to M365 Intune
--> This is where the exception happens.

**Comment by ricmestre** [comment]
@andikrueger Unfortunately I cannot share the full script, nevertheless I can do better and provide a minimal reproducible one.

On a PowerShell task this succeeds, but fails on an Azure PowerShell task (Az module is loaded by default to make a connection to Azure):

```powershell
Connect-M365Tenant `
    -ApplicationId "REDACTED" `
    -CertificateThumbprint "REDACTED" `
    -TenantId "REDACTED" `
    -Workload "MicrosoftGraph"

Get-MgGroup | Select-Object DisplayName
```

This on the other hand also fails on PowerShell task because it makes the Az module being loaded, even if I don't connect to Azure:

```powershell
Disconnect-AzAccount -ErrorAction SilentlyContinue | Out-Null

Connect-M365Tenant `
    -ApplicationId "REDACTED" `
    -CertificateThumbprint "REDACTED" `
    -TenantId "REDACTED" `
    -Workload "MicrosoftGraph"

Get-MgGroup | Select-Object DisplayName
```

Now I remember that a couple of months ago I was correctly connecting to M365 doing the following:

- Install Az.Accounts / Az.KeyVault
- Install M365Dsc
- Install M365Dsc dependencies
- Connect to Azure
- Handle certs
- Connect to M365
- Perform remaing tasks

Then I started getting this same error "You specified a different tenant - once in WithAuthority() and once using WithTenant()", for which I found a workaround on my dev tenants, and still works, that for some reason doesn't on the production and is making my brain melt, but it's definitely some update that was done on either Az module, the backend or both.

- Install Az.Accounts / Az.KeyVault
- Connect to Azure <- Connect before installing M365Dsc, but disconnecting is not required
- Install M365Dsc
- Install M365Dsc dependencies
- Handle certs
- Connect to M365
- Perform remaing tasks

**Comment by andikrueger** [comment]
Thanks for the additional details. Does this also happen in a regular PowerShell session or is it specific to a PS triggered from DevOps (Hosted Agent)?

**Comment by ricmestre** [comment]
Only on DevOps, either MShosted or through hosted agent on my VM, running the whole script directly in the VM works without issues

**Comment by ricmestre** [comment]
This is not the same issue as I reported on #3311 but is definitely a similar problem with module versions not working properly together

**Comment by andikrueger** [comment]
@peombwa do you have an idea what might be the issue here with the two modules?

**Comment by ricmestre** [comment]
Any news regarding this issue?

**Comment by NikCharlebois** [comment]
Out of curiosity, if you try connecting to the MicrosoftTeams workload using Connect-M365Tenant before trying to connect to MicrosoftGraph, do you get the error? This is not a work around, but just a way for me to troubleshoot this further. Thanks

**Comment by NikCharlebois** [comment]
I managed to reproduce the issue while walking down the authentication stack and am able to replicate at the lower levels:
![image](https://github.com/microsoft/Microsoft365DSC/assets/2547149/c0ec5d45-a260-4428-8c31-7915c3bc835d)
I reached out to the Graph team to get additional info on this.

**Comment by ricmestre** [comment]
Thank you! Much appreciated, I'll wait for your reply on this!

**Comment by NikCharlebois** [comment]
PR is opened on the Microsoft Graph PowerShell SDK side to address this:
https://github.com/microsoftgraph/msgraph-sdk-powershell/pull/2073

**Comment by NikCharlebois** [comment]
The Graph team released version 1.28.0 which will be included in our next release and which fixes this issue.
